ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,292 @@
+{
+    return [NSData dataWithBytes:certificateIdentifier length:strlen((const char *)certificateIdentifier)];
+}
+
++ (SecKeyRef) copyPrivateKeyRefForIdentifier:(NSString*)identifier
+                               correlationId:(NSUUID*)correlationId
+                                       error:(ADAuthenticationError* __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status= noErr;
+    SecKeyRef privateKeyReference = NULL;
+    
+    NSData* privateKeyTag = [NSData dataWithBytes:[identifier UTF8String] length:identifier.length];
+    
+    NSMutableDictionary* privateKeyQuery = [[NSMutableDictionary alloc] init];
+    SAFE_ARC_AUTORELEASE(privateKeyQuery);
+    
+    // Set the private key query dictionary.
+    [privateKeyQuery setObject:(__bridge id)kSecClassKey forKey:(__bridge id)kSecClass];
+    [privateKeyQuery setObject:privateKeyTag forKey:(__bridge id)kSecAttrApplicationTag];
+    [privateKeyQuery setObject:(__bridge id)kSecAttrKeyTypeRSA forKey:(__bridge id)kSecAttrKeyType];
+    [privateKeyQuery setObject:[NSNumber numberWithBool:YES] forKey:(__bridge id)kSecReturnRef];
+    
+    // Get the key.
+    status = SecItemCopyMatching((__bridge CFDictionaryRef)privateKeyQuery, (CFTypeRef*)&privateKeyReference);
+    CHECK_KEYCHAIN_STATUS(@"Failed to read WPJ private key for identifier.");
+    
+    return privateKeyReference;
+    
+_error:
+    return nil;
+}
+
++ (nullable NSString*) stringDataFromIdentifier:(nonnull NSString*)identifier
+                                  correlationId:(NSUUID*)correlationId
+                                          error:(ADAuthenticationError* __nullable __autoreleasing * __nullable)error
 {
-    (void)correlationId;
-    (void)error;
+    // Building dictionary to retrieve UPN from the keychain
+    NSMutableDictionary *query = [[NSMutableDictionary alloc] init];
+    SAFE_ARC_AUTORELEASE(query);
+    [query setObject:(__bridge id)(kSecClassGenericPassword) forKey:(__bridge id<NSCopying>)(kSecClass)];
+    [query setObject:identifier forKey:(__bridge id<NSCopying>)(kSecAttrAccount)];
+    [query setObject:(id)kCFBooleanTrue forKey:(__bridge id<NSCopying>)(kSecReturnAttributes)];
+    
+    CFDictionaryRef result = nil;
+    NSString *stringData = nil;
+    NSDictionary* resultDict = nil;
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, (CFTypeRef*)&result);
+    CHECK_KEYCHAIN_STATUS(@"String data not found for WPJ identifier");
+    
+    resultDict = (__bridge NSDictionary*)result;
+    stringData = [[resultDict objectForKey:(__bridge id)(kSecAttrService)] copy];
+    SAFE_ARC_AUTORELEASE(stringData);
+    
+    if (result)
+        CFRelease(result);
+    
+    if (!stringData || [[stringData stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] length] == 0)
+    {
+        if (error)
+            *error = [ADAuthenticationError unexpectedInternalError:@"WPJ user principal name is empty" correlationId:correlationId];
```

Comment:
The method's name is `stringDataFromIdentifier`, which means if someone else uses it for some other string this error could throw them off.